src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,906 @@
+                                              "postrouting", "nat",
+                                              NF_INET_POST_ROUTING, NF_IP_PRI_NAT_SRC + 1);
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        /* set to store ip address ranges we should masquerade for */
+        r = nft_new_set(nfnl, &batch[msgcnt], family, NFT_SYSTEMD_MASQ_SET_NAME, ++set_id, NFT_SET_INTERVAL, ip_type, ip_type_size);
+        if (r < 0)
+                goto out_unref;
+
+        /*
+         * map to store ip address:port pair to dnat to.  elements in concatenation
+         * are rounded up to 4 bytes.
+         *
+         * Example: ip protocol . tcp daddr is sizeof(uint32_t) + sizeof(uint32_t), not
+         * sizeof(uint8_t) + sizeof(uint16_t).
+         */
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        r = nft_new_map(nfnl, &batch[msgcnt], family, NFT_SYSTEMD_DNAT_MAP_NAME, ++set_id, 0,
+                        concat_types2(TYPE_INET_PROTOCOL, TYPE_INET_SERVICE), sizeof(uint32_t) * 2,
+                        concat_types2(ip_type, TYPE_INET_SERVICE), ip_type_size + sizeof(uint32_t));
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        r = sd_nfnl_message_new_dnat_rule_pre(nfnl, &batch[msgcnt], family, "prerouting");
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        r = sd_nfnl_message_new_dnat_rule_out(nfnl, &batch[msgcnt], family, "output");
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        r = sd_nfnl_message_new_masq_rule(nfnl, &batch[msgcnt], family, "postrouting");
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt < NFT_INIT_MSGS);
+        r = sd_nfnl_message_batch_end(nfnl, &batch[msgcnt]);
+        if (r < 0)
+                goto out_unref;
+
+        msgcnt++;
+        assert(msgcnt <= NFT_INIT_MSGS);
+        r = nfnl_netlink_sendv(nfnl, batch, msgcnt);
+        if (r == -EEXIST)
+                r = 0;
+
+out_unref:
+        for (i = 0; i < NFT_INIT_MSGS; i++) {
+                if (!batch[i])
+                        break;
+                sd_netlink_message_unref(batch[i]);
+        }
+
+        return r;
+}
+
+int fw_nftables_init(fw_ctx *ctx) {
+        _cleanup_(sd_netlink_unrefp) sd_netlink *nfnl = NULL;
+        int r;
+
+        r = sd_nfnl_socket_open(&nfnl);
+        if (r < 0)
+                return r;
+
+        r = fw_nftables_init_family(nfnl, AF_INET);
+        if (r < 0)
+                return r;
+
+        ctx->u.nft.nfnl = TAKE_PTR(nfnl);
+        return 0;
+}
+
+void fw_nftables_exit(fw_ctx *ctx) {
+        _cleanup_(sd_netlink_unrefp) sd_netlink *nfnl = NULL;
+
+        nfnl = TAKE_PTR(ctx->u.nft.nfnl);
```

Comment:
```c
ctx->u.ft.nfnl = sd_netlink_unref(ctx->u.ft.nfnl);
```

Reply:
changed to ctx->nfnl = sd_netlink_unref(ctx->nfnl); -- i dropped the embedded union/struct as per your other suggestion.